tts_stream_duplex.go

```diff
@@ -0,0 +1,158 @@
+package charactr
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"nhooyr.io/websocket"
+)
+
+// streamInactivityThresholdMs defines how long after inactivity the stream will be considered inactive
+const streamInactivityThresholdMs = 5000
+
+type DuplexStream struct {
+	ctx      context.Context
+	conn     *websocket.Conn
+	metadata DuplexStreamMetadata
+}
+
+type DuplexStreamMetadata struct {
+	mu               sync.Mutex
+	isClosed         bool
+	isCloseRequested bool
+	lastActiveAt     time.Time
+}
+
+// Read returns the raw audio data converted by the server
+func (v *DuplexStream) Read() ([]byte, error) {
+	messageType, audioBytes, err := v.conn.Read(v.ctx)
+	v.markStreamActivity()
+
+	// check for normal closure
+	{
+		closeErr := &websocket.CloseError{}
+		ok := errors.As(err, closeErr)
+
+		if ok && closeErr.Code == websocket.StatusNormalClosure {
+			v.metadata.mu.Lock()
+			defer v.metadata.mu.Unlock()
+			v.metadata.isClosed = true
+			return nil, io.EOF
+		}
+	}
+
+	if messageType != websocket.MessageBinary {
+		return nil, ErrStreamUnknownMessageRecv
+	}
+
+	return audioBytes, err
+}
+
+// Convert asynchronously feeds the stream with the text to be converted to audio
+func (v *DuplexStream) Convert(text string) error {
+	v.markStreamActivity()
+
+	if text == "" {
+		return ErrEmptyText
+	}
+
+	if v.metadata.isClosed || v.metadata.isCloseRequested {
+		return ErrStreamClosed
+	}
+
+	err := v.conn.Write(v.ctx, websocket.MessageText, []byte(fmt.Sprintf(`{ "type": "convert", "text": "%s" }`, text)))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Wait will block the execution until there was 5 seconds of stream inactivity
+func (v *DuplexStream) Wait() {
+	ticker := time.NewTicker(500 * time.Millisecond)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				if !v.isStreamActive() || v.metadata.isClosed || v.metadata.isCloseRequested {
+					ticker.Stop()
+					wg.Done()
+				}
+			}
+		}
+	}()
+
+	wg.Wait()
+}
+
+// Close requests the server to close the connection gracefully
+func (v *DuplexStream) Close() error {
+	v.metadata.mu.Lock()
+	defer v.metadata.mu.Unlock()
+
+	if v.metadata.isClosed || v.metadata.isCloseRequested {
+		return ErrStreamClosed
+	}
+
+	v.metadata.isCloseRequested = true
+
+	return v.conn.Write(v.ctx, websocket.MessageText, []byte(`{ "type": "close" }`))
+}
+
+// Terminate ends the stream immediately. In most cases we advise to use Close() instead.
+func (v *DuplexStream) Terminate() error {
+	v.metadata.mu.Lock()
+	defer v.metadata.mu.Unlock()
+
+	if v.metadata.isClosed || v.metadata.isCloseRequested {
+		return ErrStreamClosed
+	}
+
+	v.metadata.isClosed = true
+
+	return v.conn.Close(websocket.StatusNormalClosure, "client terminated the connection")
+}
+
+func (v *DuplexStream) authenticate(clientKey string, apiKey string) error {
+	v.markStreamActivity()
+
+	if v.metadata.isClosed || v.metadata.isCloseRequested {
```

Comment:
we need to use mutex here. How about also refactoring any metadata operations to be methods of the metadata struct?